### PR TITLE
Post Merge checkpoint sync

### LINKED ISSUE #4573: Bonsai light mode to allow checkpoint sync near head block for pivot block
For Bonsai light add a flag to do a checkpoint sync from the latest finalised block.

**Tasks**
* Choose a near head finalised block to use a target
* Override the sync target so that we only download blocks from the near head target
* Setup Teku and Besu with near-head checkpoint sync

### PR DESCRIPTION
**Pull request by gfukushima**
## PR description
Add post-merge checkpoint sync support

## Fixed Issue(s)

Fixes #4573
Linked to #4703 

## Documentation

- [x] I thought about documentation and added the `doc-change-required` label to this PR if
    [updates are required](https://wiki.hyperledger.org/display/BESU/Documentation).

## Changelog

- [ ] I thought about the changelog and included a [changelog update if required](https://wiki.hyperledger.org/display/BESU/Changelog).

### CODE REVIEW

File: besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java [hunk 1]
```diff
@@ -1761,6 +1762,7 @@ private void validateOptions() {
     validateDnsOptionsParams();
     ensureValidPeerBoundParams();
     validateRpcOptionsParams();
+    validateNearHearCheckPointSyncRequirements();
```
Comment (siladu) [inline besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java:1765]: typos:
```suggestion
    validateNearHeadCheckpointSyncRequirements();
```

File: besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java [hunk 2]
```diff
@@ -3312,6 +3320,39 @@ private void setMergeConfigOptions() {
             .isPresent());
   }
 
+  private void validateNearHearCheckPointSyncRequirements() {
```
Comment (siladu) [inline besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java:3323]: Given BesuCommand is such a huge class, is it possible to split this and the tests out into its own class?

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/SynchronizerConfiguration.java [hunk 3]
```diff
@@ -389,6 +401,11 @@ public Builder propagationManagerGetBlockTimeoutMillis(
       return this;
     }
 
+    public Builder setNearHeadCheckpointSyncEnabled(final boolean nearHeadCheckpointSyncEnabled) {
```
Comment (siladu) [inline ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/SynchronizerConfiguration.java:404]: Builder pattern convention is to omit 'set'

```suggestion
    public Builder nearHeadCheckpointSyncEnabled(final boolean nearHeadCheckpointSyncEnabled) {
```

File: besu/src/main/java/org/hyperledger/besu/controller/BesuController.java [hunk 4]
```diff
@@ -211,6 +212,13 @@ BesuControllerBuilder fromGenesisConfig(
         throw new IllegalArgumentException("Unknown consensus mechanism defined");
       }
 
+      if (configOptions.getTerminalTotalDifficulty().isPresent()
+          && configOptions.getCheckpointOptions().isValid()
+          && (UInt256.fromHexString(configOptions.getCheckpointOptions().getTotalDifficulty().get())
+              .greaterOrEqualThan(configOptions.getTerminalTotalDifficulty().get()))) {
```
Comment (siladu) [inline besu/src/main/java/org/hyperledger/besu/controller/BesuController.java:218]: Would be nice to extract this into a descriptively named method, e.g. checkpointIsLaterThanTerminalTotalDifficulty.

^ Writing that has made me wonder how a post-merge checkpoint will have a TD of >= TTD...won't all post-merge blocks have a TD = 0?

Comment (siladu) [reply besu/src/main/java/org/hyperledger/besu/controller/BesuController.java:218]: Would be good to have a test for this too I think

File: consensus/merge/src/main/java/org/hyperledger/besu/consensus/merge/PostMergeContext.java [hunk 5]
```diff
@@ -135,7 +137,9 @@ public boolean isSyncing() {
     return Optional.ofNullable(syncState.get()).map(s -> !s.isInSync()).orElse(Boolean.TRUE)
         // this is necessary for when we do not have a sync target yet, like at startup.
         // not being stopped at ttd implies we are syncing.
-        && !syncState.get().hasReachedTerminalDifficulty().orElse(Boolean.FALSE);
+        && Optional.ofNullable(syncState.get())
+            .map(s -> !s.hasReachedTerminalDifficulty().get())
+            .orElse(Boolean.FALSE);
```
Comment (siladu) [inline consensus/merge/src/main/java/org/hyperledger/besu/consensus/merge/PostMergeContext.java:142]: Good spot, looks like previous code could have got a NPE

File: consensus/merge/src/main/java/org/hyperledger/besu/consensus/merge/PostMergeContext.java [hunk 6]
```diff
@@ -265,4 +269,14 @@ private static class PayloadTuple {
       this.block = block;
     }
   }
+
+  public PostMergeContext setIsNearHeadCheckpointSync(final boolean isNearHeadCheckpointSync) {
+    this.isNearHeadCheckpointSync = isNearHeadCheckpointSync;
+    return this;
```
Comment (siladu) [inline consensus/merge/src/main/java/org/hyperledger/besu/consensus/merge/PostMergeContext.java:275]: nit: is this necessary since we're not in a builder pattern?

File: consensus/merge/src/main/java/org/hyperledger/besu/consensus/merge/PostMergeContext.java (hunk 5 above)
Comment (siladu) [reply consensus/merge/src/main/java/org/hyperledger/besu/consensus/merge/PostMergeContext.java:142]: Would be good to have a test for this too so it proves you fixed the bug.

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/engine/EngineNewPayload.java [hunk 7]
```diff
@@ -196,7 +196,8 @@ public JsonRpcResponse syncResponse(final JsonRpcRequestContext requestContext)
     }
 
     // TODO: post-merge cleanup
-    if (!mergeCoordinator.latestValidAncestorDescendsFromTerminal(newBlockHeader)) {
+    if (!mergeContext.get().isNearHeadCheckpointSync()
```
Comment (siladu) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/engine/EngineNewPayload.java:199]: unit test?

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/engine/EngineForkchoiceUpdated.java [hunk 8]
```diff
@@ -119,7 +119,8 @@ public JsonRpcResponse syncResponse(final JsonRpcRequestContext requestContext)
     }
 
     // TODO: post-merge cleanup, this should be unnecessary after merge
-    if (!mergeCoordinator.latestValidAncestorDescendsFromTerminal(newHead)) {
+    if (!mergeContext.get().isNearHeadCheckpointSync()
```
Comment (siladu) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/engine/EngineForkchoiceUpdated.java:122]: unit test?

**Review by siladu — COMMENTED** [review]
Did we decide to dynamically choose the checkpoint based on the pivot block as a separate PR or did I miss that?

File: besu/src/main/java/org/hyperledger/besu/cli/options/unstable/SynchronizerOptions.java [hunk 9]
```diff
@@ -272,6 +275,13 @@ public void parseBlockPropagationRange(final String arg) {
   private int snapsyncTrieNodeCountPerRequest =
       SnapSyncConfiguration.DEFAULT_TRIENODE_COUNT_PER_REQUEST;
 
+  @CommandLine.Option(
+      names = {NEAR_HEAD_CHECKPOINT_SYNC_FLAG},
+      hidden = true,
+      description = "Starts sync from a near-head pivot block.")
```
[the remark this answers is not in the document]
Comment (siladu) [reply besu/src/main/java/org/hyperledger/besu/cli/options/unstable/SynchronizerOptions.java:281]: Circling back - is this actually happening or do we need to manually specify the checkpoint still?

File: besu/src/main/java/org/hyperledger/besu/controller/BesuController.java [hunk 10]
```diff
@@ -211,6 +212,13 @@ BesuControllerBuilder fromGenesisConfig(
...
+        return new MergeBesuControllerBuilder().genesisConfigFile(genesisConfig);
+      }
+
       // wrap with TransitionBesuControllerBuilder if we have a terminal total difficulty:
       if (configOptions.getTerminalTotalDifficulty().isPresent()) {
```
Comment (jframe) [inline besu/src/main/java/org/hyperledger/besu/controller/BesuController.java:223]: nit: this if could become an if else as the post-merge and pre-merge are exclusive

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/engine/EngineNewPayload.java (hunk 7 above)
Comment (matkt) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/engine/EngineNewPayload.java:199]: what is the behavior when we are not using checkpoint sync ? can be strange to have a method linked to the checkpoint sync in a general class. (only my opinion)

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/engine/EngineForkchoiceUpdated.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (gfukushima) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/engine/EngineForkchoiceUpdated.java:122]: I tried that, and it was way too many changes for what this check is doing atm and IMO completely unnecessary since this is a validation that we're already planning to disable in a near future. This is only kept because some of the testnets apparently.

File: consensus/merge/src/main/java/org/hyperledger/besu/consensus/merge/PostMergeContext.java (hunk 5 above)
Comment (siladu) [reply consensus/merge/src/main/java/org/hyperledger/besu/consensus/merge/PostMergeContext.java:142]: > not sure about the .orElse(Boolean.FALSE); as we removed ! at the beginning of the line

Negation has moved into the map as far as I can see. Agree, unit test should reveal any bugs in the bug fix :)

File: besu/src/main/java/org/hyperledger/besu/controller/BesuController.java (hunk 4 above)
Comment (gfukushima) [reply besu/src/main/java/org/hyperledger/besu/controller/BesuController.java:218]: Every block after the merge has a 0 difficulty, but the total difficulty is the sum of every blocks difficulty up until the merge.
So every block after the merge has the same total difficulty which is basically 0xc70d815d562d3cfa955

File: consensus/merge/src/main/java/org/hyperledger/besu/consensus/merge/PostMergeContext.java (hunk 5 above)
Comment (matkt) [reply consensus/merge/src/main/java/org/hyperledger/besu/consensus/merge/PostMergeContext.java:142]: I wonder if changing the negation shouldn't result in an orElse(True)

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/engine/EngineNewPayload.java (hunk 7 above)
Comment (gfukushima) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/engine/EngineNewPayload.java]: Default of any other sync mode is going to have the check isNearHeadCheckpointSync returning false, to prevent it from being a blocker of this if.

Comment (gfukushima) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/engine/EngineNewPayload.java]: Also this is a very temporary piece of code that should be removed as soon as #4703 gets merged.

File: besu/src/main/java/org/hyperledger/besu/controller/BesuController.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (gfukushima) [reply besu/src/main/java/org/hyperledger/besu/controller/BesuController.java]: It is, you're right. Thanks for the heads up.

File: besu/src/test/java/org/hyperledger/besu/cli/options/SynchronizerOptionsTest.java [hunk 11]
```diff
@@ -93,7 +93,7 @@ SynchronizerOptions getOptionsFromBesuCommand(final TestBesuCommand besuCommand)
 
   @Override
   protected List<String> getFieldsToIgnore() {
-    return Arrays.asList("fastSyncMinimumPeerCount");
+    return Arrays.asList("fastSyncMinimumPeerCount", "nearHeadCheckpointSyncEnabled");
```
Comment (siladu) [inline besu/src/test/java/org/hyperledger/besu/cli/options/SynchronizerOptionsTest.java:96]: Think this should be renamed now?

File: besu/src/main/java/org/hyperledger/besu/controller/BesuController.java [hunk 12]
```diff
@@ -258,5 +263,13 @@ private Long readQbftStartBlockConfig(final QbftConfigOptions qbftConfigOptions)
 
       return startBlock;
     }
+
+    private boolean isCheckpointBlockTotalDifficultyGreaterThanTTD(
+        final GenesisConfigOptions configOptions) {
+      return configOptions.getTerminalTotalDifficulty().isPresent()
+          && configOptions.getCheckpointOptions().isValid()
+          && (UInt256.fromHexString(configOptions.getCheckpointOptions().getTotalDifficulty().get())
+              .greaterThan(configOptions.getTerminalTotalDifficulty().get()));
```
Comment (siladu) [inline besu/src/main/java/org/hyperledger/besu/controller/BesuController.java:272]: What happens if we're using a devnet or a new public testnet that starts with TTD = 0? Every blocks TD will be 0 and this condition will never be met?

File: consensus/merge/src/main/java/org/hyperledger/besu/consensus/merge/PostMergeContext.java [hunk 13]
```diff
@@ -135,7 +137,10 @@ public boolean isSyncing() {
     return Optional.ofNullable(syncState.get()).map(s -> !s.isInSync()).orElse(Boolean.TRUE)
         // this is necessary for when we do not have a sync target yet, like at startup.
         // not being stopped at ttd implies we are syncing.
-        && !syncState.get().hasReachedTerminalDifficulty().orElse(Boolean.FALSE);
+        //    && !syncState.get().hasReachedTerminalDifficulty().orElse(Boolean.FALSE);
```
Comment (siladu) [inline consensus/merge/src/main/java/org/hyperledger/besu/consensus/merge/PostMergeContext.java:140]: did you mean to leave this comment in?

File: besu/src/test/java/org/hyperledger/besu/cli/options/SynchronizerOptionsTest.java (hunk 11 above)
Comment (gfukushima) [reply besu/src/test/java/org/hyperledger/besu/cli/options/SynchronizerOptionsTest.java]: Removed actually. Thanks!

File: consensus/merge/src/main/java/org/hyperledger/besu/consensus/merge/PostMergeContext.java (hunk 13 above)
Comment (gfukushima) [reply consensus/merge/src/main/java/org/hyperledger/besu/consensus/merge/PostMergeContext.java]: Nope, thanks for catching that one

File: consensus/merge/src/main/java/org/hyperledger/besu/consensus/merge/TransitionContext.java [hunk 14]
```diff
@@ -139,4 +139,9 @@ public void putPayloadById(final PayloadIdentifier payloadId, final Block block)
   public Optional<Block> retrieveBlockById(final PayloadIdentifier payloadId) {
     return postMergeContext.retrieveBlockById(payloadId);
   }
+
+  @Override
+  public boolean isCheckpointPostMergeSync() {
+    return false;
```
Comment (jframe) [inline consensus/merge/src/main/java/org/hyperledger/besu/consensus/merge/TransitionContext.java:145]: Shouldn't this delegate to the postMergeContext.isCheckpointPostMergeSync method?

File: besu/src/main/java/org/hyperledger/besu/controller/BesuController.java [hunk 15]
```diff
@@ -258,5 +263,13 @@ private Long readQbftStartBlockConfig(final QbftConfigOptions qbftConfigOptions)
 
       return startBlock;
     }
+
+    private boolean isCheckpointBlockTotalDifficultyGreaterOrEqualThanTTD(
+        final GenesisConfigOptions configOptions) {
+      return configOptions.getTerminalTotalDifficulty().isPresent()
+          && configOptions.getCheckpointOptions().isValid()
+          && (UInt256.fromHexString(configOptions.getCheckpointOptions().getTotalDifficulty().get())
+              .greaterOrEqualThan(configOptions.getTerminalTotalDifficulty().get()));
```
Comment (siladu) [inline besu/src/main/java/org/hyperledger/besu/controller/BesuController.java:272]: Does changing this back to >= mean @jframe's point about the TransitionController being needed on the boundary case become an issue again?
https://github.com/hyperledger/besu/pull/4735#discussion_r1039180239

Maybe we need a special case for TTD = 0.

File: besu/src/test/java/org/hyperledger/besu/cli/BesuCommandTest.java [hunk 16]
```diff
@@ -5441,6 +5442,64 @@ public void portInUseReportsError() throws IOException {
...
+        "--genesis-file",
+        genesisFile.toString(),
+        "--sync-mode",
+        "X_CHECKPOINT",
+        "--Xcheckpoint-post-merge-enabled");
+
+    assertThat(commandOutput.toString(UTF_8)).isEmpty();
+    assertThat(commandErrorOutput.toString(UTF_8)).isEmpty();
+  }
+
```
Comment (siladu) [inline besu/src/test/java/org/hyperledger/besu/cli/BesuCommandTest.java:5502]: Tests for TTD < 0, TTD = 0, TTD > 0 would be good. Maybe in its own class if that means you can avoid creating load of json files?

File: consensus/merge/src/main/java/org/hyperledger/besu/consensus/merge/PostMergeContext.java [hunk 17]
```diff
@@ -135,7 +137,9 @@ public boolean isSyncing() {
     return Optional.ofNullable(syncState.get()).map(s -> !s.isInSync()).orElse(Boolean.TRUE)
         // this is necessary for when we do not have a sync target yet, like at startup.
         // not being stopped at ttd implies we are syncing.
-        && !syncState.get().hasReachedTerminalDifficulty().orElse(Boolean.FALSE);
+        && Optional.ofNullable(syncState.get())
+            .map(s -> !(s.hasReachedTerminalDifficulty().orElse(Boolean.FALSE)))
+            .orElse(Boolean.FALSE);
```
Comment (siladu) [inline consensus/merge/src/main/java/org/hyperledger/besu/consensus/merge/PostMergeContext.java:142]: There's quite a few branches of logic in this isSyncing method, are they all covered by unit tests?

I _think_ the final orElse should actually be TRUE.

If the first statement on L137 is saying "when syncState is null, then return TRUE" then I think this statement behave the same

File: besu/src/test/java/org/hyperledger/besu/cli/BesuCommandTest.java (hunk 16 above)
Comment (gfukushima) [reply besu/src/test/java/org/hyperledger/besu/cli/BesuCommandTest.java:5502]: TTD < 0? What case would we be testing here?

File: consensus/merge/src/main/java/org/hyperledger/besu/consensus/merge/PostMergeContext.java (hunk 17 above)
Comment (jframe) [reply consensus/merge/src/main/java/org/hyperledger/besu/consensus/merge/PostMergeContext.java:142]: +1

File: besu/src/test/java/org/hyperledger/besu/cli/BesuCommandTest.java (hunk 16 above)
Comment (siladu) [reply besu/src/test/java/org/hyperledger/besu/cli/BesuCommandTest.java:5502]: sorry I mistyped, I meant TD < TTD (pre-merge), TTD = TD, TD > TTD (post-merge) and finally the TD = 0 special case

File: consensus/merge/src/main/java/org/hyperledger/besu/consensus/merge/PostMergeContext.java (hunk 6 above)
Comment (gfukushima) [reply consensus/merge/src/main/java/org/hyperledger/besu/consensus/merge/PostMergeContext.java:275]: There are some methods that implement in the same way probably to make use of the method chaining feature. I've implemented this way to not break the pattern we use in the MergeBesuControllerCuilder

File: consensus/merge/src/main/java/org/hyperledger/besu/consensus/merge/PostMergeContext.java (hunk 17 above)
Comment (gfukushima) [reply consensus/merge/src/main/java/org/hyperledger/besu/consensus/merge/PostMergeContext.java]: fixed

File: besu/src/main/java/org/hyperledger/besu/controller/BesuController.java (hunk 15 above)
Comment (gfukushima) [reply besu/src/main/java/org/hyperledger/besu/controller/BesuController.java]: Added a test and another check for that case

File: besu/src/test/java/org/hyperledger/besu/cli/BesuCommandTest.java (hunk 16 above)
Comment (gfukushima) [reply besu/src/test/java/org/hyperledger/besu/cli/BesuCommandTest.java:5502]: Added some tests in the BesuController to make sure we're using the right controller in each case.

File: besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java [hunk 18]
```diff
@@ -3344,6 +3352,39 @@ private void setMergeConfigOptions() {
...
+            "Near head checkpoint sync requires a checkpoint block configured in the genesis file");
+      }
+      terminalTotalDifficulty.ifPresentOrElse(
+          value -> {
+            if (UInt256.fromHexString(
+                    genesisOptions.getCheckpointOptions().getTotalDifficulty().orElse("0x0"))
+                .lessOrEqualThan(value)) {
+              throw new InvalidConfigurationException(
+                  "Near head checkpoint sync requires a block with total difficulty greater than the TTD");
+            }
```
Comment (siladu) [inline besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java:3379]: Do we need a special case here as well for genesisTD = 0 = TTD?

Comment (siladu) [reply besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java:3379]: Have you successfully tested a post-merge genesis with this code?

**Comment by gfukushima** [comment]
Merged as #4844
